### DISCUSSION TITLE
Question about lexical scoping and rendering tables.

### DISCUSSION THREAD

**Discussion by aykaramba** [rabbibotton/clog#339] [General]
I have a div labeled "data" and I am rendering data into it.  Sometimes text, sometimes tables, sometimes other things.  For example, if I create a let* and setf a bunch of text to the div, the div will end up rendering "ddd" with this on-click event:

```
         (set-on-click btn-d
                       (lambda (obj)
                         (setf (text data) "whoa!")
                         ;; this setf sets the content of the "data" div to "whoa!" and clears out an previous test content.
                         (let* ((table (create-table data :class "w3-blue"))
                                (table-row))
                           (setf (text data) "button aaa is pressed")
                           (setf (text data) "button bbb is pressed")
                           (setf (text data) "button ccc is pressed")
                           (setf (text data) "button ddd is pressed")
```

And, if I just want to render a table to the data div, this works as well:

```          (set-on-click btn-e
                       (lambda (obj)
                         (setf (text data) "whoa!")
                         ;; this setf sets the content of the "data" div to "whoa!" and clears out an previous test content.
                         (let* ((table (create-table data :class "w3-blue"))
                                (table-row))
                           (setf table-row (create-table-row table))
                           (create-table-column table-row :content "slkdjf")
                           (create-table-column table-row :content "slkdjf")
                           (create-table-column table-row :content "slkdjf"))))
```

However, if I create a let* where I first do a bunch of setfs on the text property of the div and then try to create a table after the last "ddd" text is setf'd, the table doesn't render.  

```
         (set-on-click btn-d
                       (lambda (obj)
                         (setf (text data) "whoa!")
                         ;; this setf sets the content of the "data" div to "whoa!" and clears out an previous test content.
                         (let* ((table (create-table data :class "w3-blue"))
                                (table-row))
                           (setf (text data) "button aaa is pressed")
                           (setf (text data) "button bbb is pressed")
                           (setf (text data) "button ccc is pressed")
                           (setf (text data) "button ddd is pressed")
                           (setf table-row (create-table-row table))
                           (create-table-column table-row :content "slkdjf")
                           (create-table-column table-row :content "slkdjf")
                           (create-table-column table-row :content "slkdjf"))))
```

Why might this be?  What is going here with respect to lexical scope that is preventing the table from being rendered?  When I inspect the rendered document I only see the following:

<div class="w3-green" style="text-align:center; margin-top:40px;" id="CLOG54793">button ddd is pressed</div>

I would maybe expect a <table></table> to be rendered at minimum.

Any suggestions or tips here are helpful, this isn't an issue, I am just curious what is going on.

**Comment by rabbibotton** [comment]
The issue is that when you set the text property of a DIV it erases any child objects in the DIV, instead use the property (text-value x) and it will only change the "text" of the div and not that inner html of the div.

In general text-value is always better as it handles the correct setting of text in forms, divs, etc

**Reply by aykaramba** [reply]
Aha! Perfect, thank you very kindly.

**Comment by aykaramba** [comment]
Just a follow up on the text vs text-value in this thread.  Details:

1) I have a div referenced by a variable called "data" in a let* form.  I render tables and text to the "data" div.

2) Now that I see the difference between (setf (text-value data) "something") and (setf (text data) "something"), I have started to use (setf (text data) " ") to "clear" out the child objects in the div (by setting the content to a space effectively) and then rendering into the "data" div tables.

What was the intent of the text vs text-value?  Am I using it in a way that is in line with the design?

**Comment by rabbibotton** [comment]
I use text all the time to clear out divs etc

**Reply by aykaramba** [reply]
Thank you.
